src/db/add_package.rs

```diff
@@ -328,31 +327,103 @@ fn add_authors_into_database(
+    let crate_id = conn
+        .query("SELECT id FROM crates WHERE crates.name = $1", &[&name])?
+        .get(0)
+        .get(0);
+
+    update_owners_in_database(conn, &registry_data.owners, crate_id)?;
+
+    Ok(())
+}
+
 /// Adds owners into database
-fn add_owners_into_database(conn: &Connection, owners: &[CrateOwner], crate_id: i32) -> Result<()> {
+fn update_owners_in_database(
+    conn: &Connection,
+    owners: &[CrateOwner],
+    crate_id: i32,
+) -> Result<()> {
+    let existing_owners: Vec<String> = conn
+        .query(
+            "
+        SELECT login
+        FROM owners
+        INNER JOIN owner_rels
+            ON owner_rels.oid = owners.id
+        WHERE owner_rels.cid = $1
+    ",
+            &[&crate_id],
+        )?
+        .into_iter()
+        .map(|row| row.get(0))
+        .collect();
+
     for owner in owners {
+        debug!("Updating owner data for {}: {:?}", owner.login, owner);
+
+        // Update any existing owner data since it is mutable and could have changed since last
+        // time we pulled it
         let owner_id: i32 = {
-            let rows = conn.query("SELECT id FROM owners WHERE login = $1", &[&owner.login])?;
-            if !rows.is_empty() {
-                rows.get(0).get(0)
-            } else {
-                conn.query(
-                    "INSERT INTO owners (login, avatar, name, email)
-                                 VALUES ($1, $2, $3, $4)
-                                 RETURNING id",
-                    &[&owner.login, &owner.avatar, &owner.name, &owner.email],
-                )?
-                .get(0)
-                .get(0)
-            }
+            conn.query(
+                "
+                INSERT INTO owners (login, avatar, name, email)
+                VALUES ($1, $2, $3, $4)
+                ON CONFLICT (login) DO UPDATE
+                    SET
+                        avatar = $2,
+                        name = $3,
+                        email = $4
+                RETURNING id
+            ",
+                &[&owner.login, &owner.avatar, &owner.name, &owner.email],
+            )?
+            .get(0)
+            .get(0)
         };
 
         // add relationship
-        let _ = conn.query(
+        let updated = conn.query(
             "INSERT INTO owner_rels (cid, oid) VALUES ($1, $2)",
             &[&crate_id, &owner_id],
         );
+
+        match updated {
+            Ok(_) => debug!("Added new owner relationship"),
+            Err(e)
+                if e.as_db().and_then(|e| e.constraint.as_deref())
+                    == Some("owner_rels_cid_oid_key") =>
+            {
+                debug!("Existing owner relationship");
+            }
+            Err(e) => return Err(e.into()),
+        }
     }
+
+    let to_remove = existing_owners
+        .iter()
+        .filter(|login| !owners.iter().any(|owner| &&owner.login == login));
+
+    for login in to_remove {
+        debug!("Removing owner relationship {}", login);
```

Comment:
In https://github.com/rust-lang/docs.rs/pull/917/commits/de38887296bd4e6799ce98cf1093aa195aee67c9#diff-9297ac45637dfecac9d9fb58922d4609R66-R69 we default to no owners if an error happens. But here we remove any owners that aren't in the new list. So we could end up deleting all owners by accident. Instead, can we return a `Result` from `get_crate_owners` and don't do the update if there's an error?